src/main/java/hudson/security/LDAPSecurityRealm.java

```diff
@@ -830,50 +839,98 @@ public LdapUserDetails updateUserDetails(LdapUserDetails d) {
             synchronized (this) {
                 cached = groupDetailsCache != null ? groupDetailsCache.get(groupname) : null;
             }
             if (cached != null && cached.isValid()) {
-                cachedGroups = cached.getValue();
+                cachedGroup = cached.getValue();
             } else {
-                cachedGroups = null;
+                cachedGroup = null;
             }
         } else {
-            cachedGroups = null;
+            cachedGroup = null;
         }
 
         // TODO: obtain a DN instead so that we can obtain multiple attributes later
 
-        final Set<String> groups = cachedGroups != null
-                ? cachedGroups
-                : searchForGroupName(groupname);
-        if (cache != null && cachedGroups == null && !groups.isEmpty()) {
+        final GroupDetails group = cachedGroup != null
+                ? cachedGroup
+                : searchForGroupName(groupname, fetchMembers);
+        if (cache != null && cachedGroup == null) {
             synchronized (this) {
                 if (groupDetailsCache == null) {
-                    groupDetailsCache = new CacheMap<String, Set<String>>(cache.getSize());
+                    groupDetailsCache = new CacheMap<>(cache.getSize());
                 }
-                groupDetailsCache.put(groupname, new CacheEntry<Set<String>>(cache.getTtl(), groups));
+                groupDetailsCache.put(groupname, new CacheEntry<>(cache.getTtl(), group));
             }
         }
 
-        if(groups.isEmpty())
-            throw new UsernameNotFoundException(groupname);
-
-        return new GroupDetailsImpl(fixGroupname(groups.iterator().next()));
+        return group;
     }
 
-    private Set<String> searchForGroupName(String groupname) {
-        Set<String> groups = new TreeSet<>();
+    private @Nonnull GroupDetails searchForGroupName(String groupname, boolean fetchMembers) {
         for (LDAPConfiguration conf : configurations) {
             String searchBase = conf.getGroupSearchBase() != null ? conf.getGroupSearchBase() : "";
             String searchFilter = conf.getGroupSearchFilter() != null ? conf.getGroupSearchFilter() : GROUP_SEARCH;
-            groups.addAll(conf.getLdapTemplate().searchForSingleAttributeValues(searchBase, searchFilter, new String[]{groupname}, "cn"));
+            if (fetchMembers) {
+                return (GroupDetails)searchForFirstEntry(conf.getLdapTemplate(), searchBase, searchFilter, new String[]{groupname}, new LdapGroupDetailsMapper());
+            } else {
+                Set<String> groups = conf.getLdapTemplate().searchForSingleAttributeValues(searchBase, searchFilter, new String[]{groupname}, "cn");
+                if (!groups.isEmpty()) {
+                    return new GroupDetailsImpl(fixGroupname(groups.iterator().next()));
+                }
+            }
         }
-        return groups;
+        throw new UsernameNotFoundException(groupname);
```

Comment:
Clearer to declare even the unchecked exceptions in `throws` of this method.

Reply:
Ok, I'll update the `throws` clause.